python/lsst/cp/pipe/makeBrighterFatterKernel.py

```diff
@@ -0,0 +1,1380 @@
+                continue
+            if configDict[configParam] is True:
+                raise RuntimeError('Must set config.isr.%s to False '
+                                   'for brighter-fatter kernel calulation' % configParam)
+
+        for configParam in desirableTrue:
+            if configParam not in configDict:
+                self.log.info('Skipped enforcing %s as key was not present in config dict' % configParam)
+                continue
+            if configDict[configParam] is False:
+                self.log.warn('Found config.isr.%s set to False for brighter-fatter kernel calulation. '
+                              'It is probably desirable to have this set to True' % configParam)
+
+        # subtask settings
+        if not self.config.isr.assembleCcd.doTrim:
+            raise RuntimeError('Must trim when assembling CCDs. Set config.isr.assembleCcd.doTrim to True')
+
+    @pipeBase.timeMethod
+    def runDataRef(self, dataRef, visitPairs):
+        """Run the brighter-fatter measurement task.
+
+        For a dataRef (which is each detector here),
+        and given a list of visit pairs, calulate the
+        brighter-fatter kernel for the detector.
+
+        Parameters
+        ----------
+        dataRef : list of lsst.daf.persistence.ButlerDataRef
+            dataRef for the detector for the visits to be fit.
+        visitPairs : `iterable` of `tuple` of `int`
+            Pairs of visit numbers to be processed together
+        """
+        xcorrs = {}  # dict of lists keyed by either amp or detector depending on config.level
+        means = {}
+        kernels = {}
+
+        # setup necessary objects
+        detNum = dataRef.dataId[self.config.ccdKey]
+        if self.config.level == 'DETECTOR':
+            xcorrs = {detNum: []}
+            means = {detNum: []}
+        elif self.config.level == 'AMP':
+            # NB: don't use dataRef.get('raw_detector')
+            # this currently doesn't work for composites because of the way
+            # composite objects (i.e. LSST images) are handled/constructed
+            # these need to be retrieved from the camera and dereferenced
+            # rather than accessed directly
+            detector = dataRef.get('camera')[dataRef.dataId[self.config.ccdKey]]
+            ampInfoCat = detector.getAmpInfoCatalog()
+            ampNames = [amp.getName() for amp in ampInfoCat]
+            xcorrs = {key: [] for key in ampNames}
+            means = {key: [] for key in ampNames}
+        else:
+            raise RuntimeError("Unsupported level: {}".format(self.config.level))
+
+        # calculate or retrieve the gains
+        if self.config.doCalcGains:
+            self.log.info('Beginning gain estimation for detector %s' % detNum)
+            gains, nomGains = self.estimateGains(dataRef, visitPairs)
+            dataRef.put(gains, datasetType='brighterFatterGain')
+            self.log.info('Finished gain estimation for detector %s' % detNum)
```

Comment:
Consider making this debug level. I worry about users complaining about excessive verbosity. In which case the "beginning" log message may want to be "Compute gain estimation for detector %s"